packages/@react-stately/datepicker/src/useDateFieldState.ts

```diff
@@ -234,15 +236,19 @@ export function useDateFieldState<T extends DateValue = DateValue>(props: DateFi
     if (props.isDisabled || props.isReadOnly) {
       return;
     }
+    let validKeys = Object.keys(validSegments);
+    let allKeys = Object.keys(allSegments);
 
-    if (Object.keys(validSegments).length >= Object.keys(allSegments).length) {
+    // if all the segments are completed or a timefield with everything but am/pm set the time, also ignore when am/pm cleared
+    if (Object.keys(validSegments).length >= Object.keys(allSegments).length || (validKeys.length === allKeys.length - 1 && allSegments.dayPeriod && !validSegments.dayPeriod && clearedSegment !== 'dayPeriod')) {
```

Comment:
Unmounting with a ref didn't work because it was updating on segment changes and when the dialog dismisses the onChange happens before the component unmounts.